### ISSUE TITLE
Ephemeral Datasets/Snapshots

### ISSUE THREAD

**Issue by ngie-eign** [openzfs/zfs#14956]
### Describe the feature would like to see added to OpenZFS

As a consumer of ZFS on FreeBSD, I have written a python package to help manage a series of rotating snapshots on a schedule (source can be found [here](https://github.com/ngie-eign/scratch/tree/main/profiles/base/freebsd/root/bin/zfs_snapshot)).

This script is run by cron on a periodic basis (hourly, daily, etc), and handles deletion of snapshots with a predefined (by script invocation) policy:
```
@hourly         zfs_snapshot --snapshot-period hours --recursive
@daily          zfs_snapshot --snapshot-period days --recursive
@monthly        zfs_snapshot --snapshot-period months --recursive
```

What is missing and what would simplify my life is two items:
1. A native snapshot subsystem/daemon.
2. A means for snapshots to be tagged with properties which allow it to auto-expire based on a time period, such that my the core logic of my script which deals with time computation, etc, doesn't need to exist: all that would need to be done is at snapshot creation time (or afterwards) a property could be set on the snapshot for it to automatically be deleted within a specific timeframe. If the property was not set or was removed, the snapshot would be treated like any other snapshot.

### How will this feature improve OpenZFS?

- Data protection is important.
- Enterprise storage solutions today are large, but still finite. With smaller deployments, like mine it's felt more acutely. Redundant/unnecessary data wastes space on my system requiring preening (automatic or manual) to be done from time to time.
- Snapshots captured between 2 fixed points in time, e.g., A and B, doesn't matter as much to some system administrators as the data as presented at times A and B.

In this day and age with data holds (legal) and other legal requirements (GDPR) around data retention, more system administrators and integrators have to design systems which ensure that data lives for a specific period of time, but sometimes no longer than a another point in time. When I worked at a large social media company, I helped design one of those systems to deal with things like GDPR.

With climate change being an increasingly important issue, along with waste from managing large datasets, such as AI/ML, combined with the amount of ephemeral data as it exists in today's big data world, having a system in ZFS, etc, which better integrates and manages ephemeral data such that data holds and automatic deletion of data are more hands-off would be a huge win to Big Data/Data Lake administrators, as well as smaller NAS system integrators, like myself.

### Additional context

I can provide more details on request.

**Comment by ngie-eign** [comment]
It seems that `zfs hold`/`zfs release` (a feature I wasn't aware of) would help fill some of these gaps, but extending the "hold" concept to be more flexible and adding date attributes to snapshots and making it sortable/filterable so companion daemons/utilities in userspace could help manage the data, e.g., hold and automatically purge it if need be and when appropriate.

**Comment by GregorKopka** [comment]
You can easily script that by setting a user property with the timestamp where the snapshot should be pruned, then (given you used _expire_ as the property) do `zfs list -o name,_expire_ -s _expire_ -H -p`to get a list that you iterate over to destroy the snapshots where _expire_ < now().

A single line of bash can reach what you want and can be executed manually or from cron.

As a not-so-technical comment regarding the last section:
Please keep unrelated stuff, like climate change fear-porn (the _man made panic_ about perfectly normal natural warming at the end of the current ice age), out of technical discussions.

**Comment by chrisrd** [comment]
Right back atcha @GregorKopka : please keep unrelated stuff, like science denialist nonsense, out of technical discussions. (Not that I want an argument, but please feel free to contact me privately if you'd like to address this further.)

Perhaps we'd best leave it at "please keep unrelated stuff, especially anything likely to have disparate strong viewpoints, out of technical discussions". These points from the [OpenZFS Code of Conduct](https://openzfs.org/wiki/Code_of_Conduct) could provide guidance on this issue:

- Be friendly and patient.
- Be welcoming.
- Be considerate.
- Be respectful.
- Be careful in the words that you choose.

**Comment by GregorKopka** [comment]
My friendliness and respectfulness find themselves under attack when they encounter attempts to doctor others into supporting (or even to work for) something, _because else we'll all die of what we're currently been told (by the powers that be) to die of, unless we give up our basic human rights_. I will always object such attacks on the minds of humans.

Shall we directly open cases at the _Working group of Code of Conduct mechanics_ about such disrespectful attempts at manipulation, instead of pointing them out directly (and, still, politely)? Because I have no idea how else the _Guideline_ of 
> Repeated harassment of others. In general, if someone asks you to stop, then stop.

should work, if it is not possible _to ask someone to stop_.

Also, he got a working answer for his problem, nevertheless.

**Comment by ngie-eign** [comment]
> My...

Please stop playing the victim. It's not a good look.
 
> Also, he got a working answer for his problem, nevertheless.

No. Your suggested workaround does not address my ask. It requires the system administrator to supply information which the tool/file system could itself provide/expose to consumers, but doesn't today.

Providing it in an unstructured way means you will have to deal with a localization/data sanitization problem. Moreover, you will need to come up with a way to "uniqueify" snapshots, given that the datas that my script bakes in effectively do that today. Finally, you will need to come up with a way to scrub the data after the fact in order to adhere to strict rules and norms around data retention (see GDPR).

Look at it the problem this way, if you view it from a capitalist lens: data/storage waste is becoming an increasingly important problem which is requiring more servers and more data center space in order to facilitative was is effectively "information hoarding". If this wasn't an issue, I wouldn't be working in the industry I am today. Moreover, legal holds have large financial risks associated with them: if companies don't dispose of the data properly (including shredding hard media), they could be sued for malfeasance.

Oh, and I don't go by "he", bro.

**Comment by chrisrd** [comment]
Hi Enji (and Gregor),

On Thu, Jun 29, 2023 at 08:01:48PM -0700, Enji Cooper wrote:
> Please stop playing the victim. It's not a good look.

Please, let's all take this down a notch and recognise that people here 
have different backgrounds and sensitivities: that's why it's better to 
keep sensitive subjects out of the technical discussions.

>> Also, he got a working answer for his problem, nevertheless.
>
> No. Your suggested workaround does not address my ask. It requires the 
> system administrator to supply information which the tool/file system 
> could itself provide/expose to consumers, but doesn't today.
>
> Providing it in an unstructured way means you will have to deal with a 
> localization/data sanitization problem. Moreover, you will need to come 
> up with a way to "uniqueify" snapshots, given that the datas that my 
> script bakes in effectively do that today. Finally, you will need to 
> come up with a way to scrub the data after the fact in order to adhere 
> to strict rules and norms around data retention (see GDPR).

I actually thought Gregor's suggestion was quite appropriate on this (and 
probably should have said so in my own somewhat snarky response - 
apologies to you both). It's basically how I handle it in my own rotation 
scripts.

This indicates I and Gregor maybe don't quite understand your request: the 
issue as I understand it is that you'd like snapshots to be able to 
"auto-expire based on a time period". The solution we've all come up with 
is to run a userland script that removes snaphots based on some locally 
defined retention and rotation policy. As Gregor pointed out that can be 
an almost trivial task (of course for complex policies it may be anything 
but trivial).

What more are you looking for? You've mentioned missing "a native snapshot 
subsystem/daemon". Are you simply wanting to replace your own scripts with 
such a built-in facility?

One problem I see is there are many different ways of scheduling 
rotations, I'm not sure there's a one size fits all scheduling mechanism 
that would fit, without getting overly complex to the point where 
configuring the native daemon might be nearly as big a task as writing a 
simple script in the first place. On the other hand, there's probably only 
a few ways that most people would schedule such things, so perhaps a 
simple option might cover 80-90% of the use cases out there, and anyone 
with more complex needs could still script their own. I think that would 
actually be a help for many people so we don't all have to come up with 
our own scripts over and over again.

I'm not subject to GDPR so haven't considered "scrubbing the data after 
the fact". Wouldn't encrypting the disks do enough to satisfy GDPR, in the 
sense that a removed disk would not have any visible data on it? Or are 
you wanting to write zeros / random data to the parts of the media that 
are no longer referenced by the snapshots so even an attacker with root 
privileges can't get at that old data?

Cheers,

Chris

**Comment by GregorKopka** [comment]
We can learn from `zfs sharesmb` that one-size-fits-all and baked-in is an approach that makes more problems than it solves: While it gives an easy OOB way to share datasets using samba, it completely and miserably fails the moment you eg. need to add additional options to the user share (like access control) that have not been foreseen, resulting in not only having to implement what you actually need anyway, but also to code around any built-in magic that not only does not work for but very likely also gets in the way of what you try to achieve.

There also is "Those who don't understand UNIX are doomed to reinvent it, poorly."
--Henry Spencer

All parts you need to solve your _specialized business case_ (_cron and a scripting language of your choice_) exist on a system that can run ZFS, only thing that is missing is _someone to actually code it_. Offloading that through a feature request here brings, from my perspective, no benefit to OpenZFS.

**Comment by ngie-eign** [comment]
I apologize for the tone in my previous response: I've been frustrated with other things and vented here (which was the wrong forum).
Thinking more out loud, yes, OpenZFS is the wrong place for this feature request: it's better suited for ZFS integrators like FreeNAS, Nexenta, etc, who can implement end-to-end features like this.
I'll open up some targeted issues for missing pieces that could be implemented in OpenZFS to avoid the need for reinventing the wheel with this feature (and also my limited ZFS use case):
1. Exposing the "hold" property of a given dataset.
2. Providing structured output from `zpool list` and `zfs list`, e.g., using CSV, etc, so sysadmin tasks could be scripted more easily. Shells aren't capable of doing this task at scale -- parsing and acting on the data would need to be done in a higher-level scripting language like go, python, ruby, etc.

**Comment by ngie-eign** [comment]
Also: the GDPR/legal hold scrubbing issue is tricky. I used to work for a large company that wiped out data, but because of uncertainty in terms of whether or not it had been scrubbed properly, sent disks to a physical shredder. I don't know if they do it still, but there are groups that don't trust that data has been overwritten a handful of times by random data (which I was told was "good enough for the NSA" at other jobs). The big company's CYOA stance was that the only way the data couldn't be reconstituted is through physical destruction, so physical destruction it was..

**Comment by GregorKopka** [comment]
> 1. Exposing the "hold" property of a given dataset.

Currently a hold can only be set on a snapshots and `zfs holds` expects a single snapshot to operate on.
Relaxing the requirement of the latter, so it could report on all snapshots of a dataset (and possibly children, ideally selected through `-r / -d` to recycle the interface of `zfs list`) would IMHO be a good improvement.

>  2. Providing structured output from `zpool list` and `zfs list`, e.g., using CSV, etc, so sysadmin tasks could be scripted more easily. Shells aren't capable of doing this task at scale -- parsing and acting on the data would need to be done in a higher-level scripting language like go, python, ruby, etc.

You can already get _kind-of_ CSV with `-H -p`:
-H separates fields by **a single tab** as the separator
-p gives easily parsable numbers (like timestamps instead of human readable dates)
Tab as separator has the benefit over `,` (comma) that the latter is valid content for property values (which would mess up CSV when querying multiple properties through `zfs list`) and most shells have a way to split a read line into an array.

For bash scripts you can use the following approach to get to the information you need:
`zfs list -H  | while read LINE; do arr=(${LINE//\\t/ }); [...]; done`
just add -d -p -o -s flags to the `zfs list` as needed and access the resulting array inside the loop with `${arr[x]}` (with x being the index which field you want).

Would you use `-t snap -o name,expiration -p` and compare ${arr[1]} against `date +%s` to decide if you have to `zfs destroy ${arr[0]}`... you would have solved the original issue.

Bottom line: 1. sounds good, 2. would be redundant, apology accepted.
